src/java.base/share/classes/java/io/ObjectInputStream.java

```diff
@@ -2275,8 +2283,26 @@ private Object readOrdinaryObject(boolean unshared)
             if (!unshared)
                 handles.setObject(passHandle, obj);
         } else if (desc.isExternalizable()) {
+            assert obj != null;
+            if (desc.isValue()) {
+                throw new NotSerializableException("Externalizable not valid for value class "
+                        + cl.getName());
+            }
+            if (!unshared)
+                handles.setObject(passHandle, obj);
             readExternalData((Externalizable) obj, desc);
+        } else if (desc.isValue()) {
+            // For value objects, read the fields and finish the buffer before publishing the ref
+            assert obj != null;
+            readSerialData(obj, desc);
+            obj = desc.finishValue(obj);
+            if (!unshared)
+                handles.setObject(passHandle, obj);
         } else {
+            // For all other objects, publish the ref and then read the data
+            assert obj != null;
+            if (!unshared)
+                handles.setObject(passHandle, obj);
             readSerialData(obj, desc);
         }
```

Comment:
These `assert obj != null` statements can be hoisted before all the `if`‑statements.

Reply:
Except the first, for Records object must be null.